File: zul/src/archive/web/js/zul/wgt/Notification.js
/* Notification.js

	Purpose:
		
	Description:
		
	History:
		Thu Mar 15 17:12:46     2012, Created by simon

Copyright (C) 2012 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
(function () {
	var _iconMap = {
		'warning': 'z-icon-exclamation-sign',
		'info': 'z-icon-info-sign',
		'error': 'z-icon-remove-sign'
	};
	var _dirMap = {
		'u': 'up',
		'd': 'down',
		'l': 'left',
		'r': 'right'
	};
/**
 * A notification widget.
 * @since 6.0.1
 */
zul.wgt.Notification = zk.$extends(zul.wgt.Popup, {
	
	$init: function (msg, opts) {
		this.$supers(zul.wgt.Notification, '$init', arguments);
		this._msg = msg;
		this._type = opts.type;
		this._ref = opts.ref;
		this._dur = opts.dur;
		this._closable = opts.closable;
	},
	redraw: function (out) {
		var uuid = this.uuid,
			icon = this.$s('icon');
		out.push('<div', this.domAttrs_(), '>');
		if (this._ref) //output arrow if reference exist
			out.push('<div id="', uuid, '-p" class="', this.$s('pointer'), '"></div>');
		out.push('<i id="', uuid, '-icon" class="', icon, ' ', (_iconMap[this._type]), '"></i>');
		out.push('<div id="', uuid, '-cave" class="', this.$s('content'), '">',
				this._msg, '</div>');
		if (this._closable)
			out.push('<div id="', uuid, '-cls" class="', this.$s('close'),
					'"><i class="', icon, ' z-icon-remove"></i></div>');
		out.push('</div>'); // not encoded to support HTML
	},
	domClass_: function (no) {
		var type = this._type,
			ref = this._ref,
			s = this.$supers(zul.wgt.Notification, 'domClass_', arguments);
		if (type)
			s += ' ' + this.$s(zUtl.encodeXML(type));
		return s;
	},
	doClick_: function (evt) {
		var p = evt.domTarget;
		if (jq.contains(this.$n('cls'), p))
			this.close();
		else
			this.$supers('doClick_', arguments);
	},
	onFloatUp: function(ctl, opts) {
		if (opts && opts.triggerByFocus) //only mouse click should close notification
			return;
		if (!this.isVisible())
			return;
		var wgt = ctl.origin;
		for (var floatFound; wgt; wgt = wgt.parent) {
			if (wgt == this) {
				if (!floatFound) 
					this.setTopmost();
				return;
			}
			if (wgt == this.parent && wgt.ignoreDescendantFloatUp_(this))
				return;
			floatFound = floatFound || wgt.isFloating_();
		}
		if (!this._closable && this._dur <= 0)
			this.close({sendOnOpen:true});
	},
	open: function (ref, offset, position, opts) {
		this.$supers(zul.wgt.Notification, 'open', arguments);
		this._fixarrow(ref); //ZK-1583: modify arrow position based on reference component
		zk(this).redoCSS(-1, {'fixFontIcon': true});
	},
	position: function (ref, offset, position, opts) {
		this.$supers(zul.wgt.Notification, 'position', arguments);
		this._fixarrow(ref); //ZK-1583: modify arrow position based on reference component
	},
	_posInfo: function (ref, offset, position, opts) {
		this._fixPadding(position);
		return this.$supers(zul.wgt.Notification, '_posInfo', arguments);
	},
	_fixPadding: function (position) {
		var p = this.$n('p');
		if (!p)
			return;
		var n = this.$n(),
			pw = 2 + (zk(p).borderWidth() / 2) || 0,
			ph = 2 + (zk(p).borderHeight() / 2) || 0;
		
		n.style.padding = '0';
		// cache arrow direction for _fixarrow() later
		switch (position) {
		case "before_start":
		case "before_center":
		case "before_end":
			this._dir = 'd';
			n.style.paddingBottom = ph + 'px';
			break;
		case "after_start":
		case "after_center":
		case "after_end":
			this._dir = 'u';
			n.style.paddingTop = ph + 'px';
			break;
		case "end_before":
		case "end_center":
		case "end_after":
			this._dir = 'l';
			n.style.paddingLeft = pw + 'px';
			break;
		case "start_before":
		case "start_center":
		case "start_after":
			this._dir = 'r';
			n.style.paddingRight = pw + 'px';
			break;
		case "top_left":
		case "top_center":
		case "top_right":
		case "middle_left":
		case "middle_center":
		case "middle_right":
		case "bottom_left":
		case "bottom_center":
		case "bottom_right":
		case "overlap":
		case "overlap_end":
		case "overlap_before":
		case "overlap_after":
			this._dir = 'n';
			break;
		// at_pointer, after_pointer, etc.
		default:
			this._dir = 'n';
		}
	},
	_fixarrow: function (ref) {
		var p = this.$n('p');
		if (!p)
			return;
		
		var pzcls = this.$s('pointer'),
			n = this.$n(),
			refn = ref.$n(),
			dir = this._dir,
			zkp = zk(p),
			pw = zkp.borderWidth(),
			ph = zkp.borderHeight(),
			nOffset = zk(n).cmOffset(),
			refOffset = zk(refn).cmOffset(),
			arrXOffset = (refn.offsetWidth - pw) / 2,
			arrYOffset = (refn.offsetHeight - ph) / 2;
		if (dir != 'n') {
			// positioning
			if (dir == 'u' || dir == 'd') {
				var b = dir == 'u',
					l1 = (n.offsetWidth - pw) / 2 || 0,
					l2 = refOffset[0] - nOffset[0] + arrXOffset || 0;
				p.style.left = (refn.offsetWidth >= n.offsetWidth ? l1 : l2) + 'px'; //ZK-1583: assign arrow position to reference widget if it is smaller than notification
				p.style[b ? 'top' : 'bottom'] = ((2 - ph / 2) || 0) + 'px';
				p.style[b ? 'bottom' : 'top'] = '';
			} else {
				var b = dir == 'l',
					t1 = (n.offsetHeight - ph) / 2 || 0,
					t2 = refOffset[1] - nOffset[1] + arrYOffset || 0;
				p.style.top = (refn.offsetHeight >= n.offsetHeight ? t1 : t2) + 'px'; //ZK-1583: assign arrow position to reference widget if it is smaller than notification
				p.style[b ? 'left' : 'right'] = ((2 - pw / 2) || 0) + 'px';
				p.style[b ? 'right' : 'left'] = '';
			}
			
			p.className = pzcls + (_dirMap[dir] ? ' ' + this.$s(_dirMap[dir]) : '');
			jq(p).show();
			
		} else {
			p.className = pzcls;
			jq(p).hide();
		}
	},
	openAnima_: function (ref, offset, position, opts) {
		var self = this;
		jq(this.$n()).fadeIn(500, function () {
			self.afterOpenAnima_(ref, offset, position, opts);
		});
	},
	closeAnima_: function (opts) {
		var self = this;
		jq(this.$n()).fadeOut(500, function () {
			self.afterCloseAnima_(opts);
		});
	},
	afterCloseAnima_: function (opts) {
		this.detach();
	}
}, {
	
	/**
	 * Shows a notification.
	 * TODO
	 */
	show: function (msg, pid, opts) {
		if (!opts)
			opts = {};
		var parent = zk.Widget.$(pid),
			ref = opts.ref,
			pos = opts.pos,
			dur = opts.dur,
			ntf = new zul.wgt.Notification(msg, opts),
			off = opts.off;
		
		// TODO: allow user to specify arrow direction?
		
		if (!pos && !off)
			pos = ref ? "end_center" : "middle_center";
		
		if (!parent) {
			// bug ZK-1136: If target page is detached, append to current active page
			parent = zk.Desktop.$().firstChild;
		}
		parent.appendChild(ntf);
		ntf.open(ref, off, pos);
		
		// auto dismiss
		if (dur > 0)
			setTimeout(function () {
				if (ntf.desktop)
					ntf.close();
			}, dur);
	}
	
});

})();

File: zk/src/main/java/org/zkoss/zk/ui/util/Notification.java
/* Notification.java

	Purpose:
		
	Description:
		
	History:
		Mon Sep 16 17:22:31 CST 2019, Created by rudyhuang

Copyright (C) 2019 Potix Corporation. All Rights Reserved.
*/
package org.zkoss.zk.ui.util;

import org.zkoss.zk.au.out.AuNotification;
import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.Execution;
import org.zkoss.zk.ui.Executions;
import org.zkoss.zk.ui.Page;
import org.zkoss.zk.ui.sys.ExecutionCtrl;

/**
 * Utilities to send Notification to the client.
 *
 * Note: this class makes accessing 'client-side' more directly.
 * Methods such as {@link #show(String)} will not encode the strings passed into them,
 * thus the formatting of messages at 'client-side' is allowed.
 * User input should be escaped carefully.
 *
 * @author rudyhuang
 * @since 9.0.0
 */
public class Notification {
	/**
	 * Notification type: information
	 */
	public static final String TYPE_INFO = "info";

	/**
	 * Notification type: warning
	 */
	public static final String TYPE_WARNING = "warning";

	/**
	 * Notification type: error
	 */
	public static final String TYPE_ERROR = "error";


	/**
	 * Displays a message.
	 *
	 * @param msg the message to show
	 * @param type available types are "info", "warning", "error"
	 * @param ref the referenced component, null to be based on browser window
	 * @param position predefined positions.
	 * <p> Available options are:
	 * <ul>
	 *  <li><b>before_start</b><br/> the message appears above the anchor, aligned to the left.</li>
	 *  <li><b>before_center</b><br/> the message appears above the anchor, aligned to the center.</li>
	 *  <li><b>before_end</b><br/> the message appears above the anchor, aligned to the right.</li>
	 *  <li><b>after_start</b><br/> the message appears below the anchor, aligned to the left.</li>
	 *  <li><b>after_center</b><br/> the message appears below the anchor, aligned to the center.</li>
	 *  <li><b>after_end</b><br/> the message appears below the anchor, aligned to the right.</li>
	 *  <li><b>start_before</b><br/> the message appears to the left of the anchor, aligned to the top.</li>
	 *  <li><b>start_center</b><br/> the message appears to the left of the anchor, aligned to the middle.</li>
	 *  <li><b>start_after</b><br/> the message appears to the left of the anchor, aligned to the bottom.</li>
	 *  <li><b>end_before</b><br/> the message appears to the right of the anchor, aligned to the top.</li>
	 *  <li><b>end_center</b><br/> the message appears to the right of the anchor, aligned to the middle.</li>
	 *  <li><b>end_after</b><br/> the message appears to the right of the anchor, aligned to the bottom.</li>
	 *  <li><b>overlap/top_left</b><br/> the message overlaps the anchor, with anchor and message aligned at top-left.</li>
	 *  <li><b>top_center</b><br/> the message overlaps the anchor, with anchor and message aligned at top-center.</li>
	 *  <li><b>overlap_end/top_right</b><br/> the message overlaps the anchor, with anchor and message aligned at top-right.</li>
	 *  <li><b>middle_left</b><br/> the message overlaps the anchor, with anchor and message aligned at middle-left.</li>
	 *  <li><b>middle_center</b><br/> the message overlaps the anchor, with anchor and message aligned at middle-center.</li>
	 *  <li><b>middle_right</b><br/> the message overlaps the anchor, with anchor and message aligned at middle-right.</li>
	 *  <li><b>overlap_before/bottom_left</b><br/> the message overlaps the anchor, with anchor and message aligned at bottom-left.</li>
	 *  <li><b>bottom_center</b><br/> the message overlaps the anchor, with anchor and message aligned at bottom-center.</li>
	 *  <li><b>overlap_after/bottom_right</b><br/> the message overlaps the anchor, with anchor and message aligned at bottom-right.</li>
	 *  <li><b>at_pointer</b><br/> the message appears with the upper-left aligned with the mouse cursor.</li>
	 *  <li><b>after_pointer</b><br/> the message appears with the top aligned with
	 *  	the bottom of the mouse cursor, with the left side of the message at the horizontal position of the mouse cursor.</li>
	 * </ul></p>
	 * @param duration the duration of notification in millisecond. If zero or
	 * negative the notification does not dismiss until user left-clicks outside
	 * of the notification box.
	 * @param closable whether to close notification manually or not. If true there will be a
	 * close button on notification message and won't close until user click the button
	 * or duration time up, default false.
	 */
	public static void show(String msg, String type, Component ref, String position, int duration, boolean closable) {
		Execution exec = Executions.getCurrent();
		Page page = ref != null ? ref.getPage() : null;
		if (page == null && exec instanceof ExecutionCtrl)
			page = ((ExecutionCtrl) exec).getCurrentPage();
		if (type == null)
			type = TYPE_INFO;
		exec.addAuResponse(new AuNotification(msg, type, page, ref, position, duration, closable));
	}

	/**
	 * Displays a message.
	 * @param msg the message to show
	 * @param type available types are "info", "warning", "error"
	 * @param ref the referenced component, null to be based on browser window
	 * @param x the horizontal position of the notification, aligned at top-left (in pixel)
	 * @param y the vertical position of the notification, aligned at top-left (in pixel)
	 * @param duration the duration of notification in millisecond. If zero or
	 * negative the notification does not dismiss until user left-clicks outside
	 * of the notification box.
	 * @param closable whether to close notification manually or not. If true there will be a
	 * close button on notification message and won't close until user click the button
	 * or duration time up, default false.
	 */
	public static void show(String msg, String type, Component ref, int x, int y, int duration,
	                        boolean closable) {
		Execution exec = Executions.getCurrent();
		Page page = ref != null ? ref.getPage() : null;
		if (page == null && exec instanceof ExecutionCtrl)
			page = ((ExecutionCtrl) exec).getCurrentPage();
		if (type == null)
			type = TYPE_INFO;
		exec.addAuResponse(new AuNotification(msg, type, page, ref, x, y, duration, closable));
	}

	/**
	 * Displays a message.
	 * @param msg the message to show
	 * @param type available types are "info", "warning", "error"
	 * @param ref the referenced component, null to be based on browser window
	 * @param x the horizontal position of the notification, aligned at top-left (in pixel)
	 * @param y the vertical position of the notification, aligned at top-left (in pixel)
	 * @param duration the duration of notification in millisecond. If zero or
	 * negative the notification does not dismiss until user left-clicks outside
	 * of the notification box.
	 */
	public static void show(String msg, String type, Component ref, int x, int y, int duration) {
		show(msg, type, ref, x, y, duration, false);
	}

	/**
	 * Displays a message.
	 * @param msg the message to show
	 * @param type available types are "info", "warning", "error"
	 * @param ref the referenced component, null to be based on browser window
	 * @param position predefined positions.
	 * <p> Available options are:
	 * <ul>
	 *  <li><b>before_start</b><br/> the message appears above the anchor, aligned to the left.</li>
	 *  <li><b>before_center</b><br/> the message appears above the anchor, aligned to the center.</li>
	 *  <li><b>before_end</b><br/> the message appears above the anchor, aligned to the right.</li>
	 *  <li><b>after_start</b><br/> the message appears below the anchor, aligned to the left.</li>
	 *  <li><b>after_center</b><br/> the message appears below the anchor, aligned to the center.</li>
	 *  <li><b>after_end</b><br/> the message appears below the anchor, aligned to the right.</li>
	 *  <li><b>start_before</b><br/> the message appears to the left of the anchor, aligned to the top.</li>
	 *  <li><b>start_center</b><br/> the message appears to the left of the anchor, aligned to the middle.</li>
	 *  <li><b>start_after</b><br/> the message appears to the left of the anchor, aligned to the bottom.</li>
	 *  <li><b>end_before</b><br/> the message appears to the right of the anchor, aligned to the top.</li>
	 *  <li><b>end_center</b><br/> the message appears to the right of the anchor, aligned to the middle.</li>
	 *  <li><b>end_after</b><br/> the message appears to the right of the anchor, aligned to the bottom.</li>
	 *  <li><b>overlap/top_left</b><br/> the message overlaps the anchor, with anchor and message aligned at top-left.</li>
	 *  <li><b>top_center</b><br/> the message overlaps the anchor, with anchor and message aligned at top-center.</li>
	 *  <li><b>overlap_end/top_right</b><br/> the message overlaps the anchor, with anchor and message aligned at top-right.</li>
	 *  <li><b>middle_left</b><br/> the message overlaps the anchor, with anchor and message aligned at middle-left.</li>
	 *  <li><b>middle_center</b><br/> the message overlaps the anchor, with anchor and message aligned at middle-center.</li>
	 *  <li><b>middle_right</b><br/> the message overlaps the anchor, with anchor and message aligned at middle-right.</li>
	 *  <li><b>overlap_before/bottom_left</b><br/> the message overlaps the anchor, with anchor and message aligned at bottom-left.</li>
	 *  <li><b>bottom_center</b><br/> the message overlaps the anchor, with anchor and message aligned at bottom-center.</li>
	 *  <li><b>overlap_after/bottom_right</b><br/> the message overlaps the anchor, with anchor and message aligned at bottom-right.</li>
	 *  <li><b>at_pointer</b><br/> the message appears with the upper-left aligned with the mouse cursor.</li>
	 *  <li><b>after_pointer</b><br/> the message appears with the top aligned with
	 *  	the bottom of the mouse cursor, with the left side of the message at the horizontal position of the mouse cursor.</li>
	 * </ul></p>
	 * @param duration the duration of notification in millisecond. If zero or
	 * negative the notification does not dismiss until user left-clicks outside
	 * of the notification box.
	 */
	public static void show(String msg, String type, Component ref, String position, int duration) {
		show(msg, type, ref, position, duration, false);
	}

	/**
	 * Shows a message at the right side of the given component.
	 * @param msg the message to show
	 * @param ref the referenced component, null to be based on browser window
	 * @param closable whether to close notification manually or not. If true there will be a
	 * close button on notification message and won't close until user click the button
	 * or duration time up, default false.
	 */
	public static void show(String msg, Component ref, boolean closable) {
		show(msg, null, ref, null, -1, closable);
	}

	/**
	 * Shows a message at the right side of the given component.
	 * @param msg the message to show
	 * @param ref the referenced component, null to be based on browser window
	 */
	public static void show(String msg, Component ref) {
		show(msg, null, ref, null, -1, false);
	}

	/**
	 * Shows a message at the center of the browser window.
	 * @param msg the message to show
	 */
	public static void show(String msg) {
		show(msg, null, null, null, -1, false);
	}

	/**
	 * Shows a message at the center of the browser window.
	 * @param msg the message to show
	 * @param closable whether to close notification manually or not. If true there will be a
	 * close button on notification message and won't close until user click the button
	 * or duration time up, default false.
	 */
	public static void show(String msg, boolean closable) {
		show(msg, null, null, null, -1, closable);
	}
}
